# 🐍 Makine Öğrenmesi — Python Lab 02
## Parametrik yöntemler ve çok değişkenli yöntemlere giriş

**HOCA VERSİYONU · ÇÖZÜMLÜ · 2. HAFTA**  
**Hedef kitle:** Python ile ilk tanışmasını geçen hafta yapmış öğrenciler  
**Ders süresi:** 90 dakika (75 dakikaya indirme önerisi en sonda)  
**Gerekenler:** Jupyter Notebook / Google Colab, Python 3. **Harici kütüphane yok.**

### Bu laboratuvarda ne kazanacağız?
- Geçen haftaki liste, `for`, `if` ve değişken bilgilerini gerçek veri üzerinde kullanmak.
- Ortalama, varyans ve standart sapmayı hesaplamak.
- Bir veri kümesinden **gözlenen (ampirik) olasılık** elde etmek.
- Vektör, matris benzeri veri düzeni ve skaler çarpımı kodlamak.
- `y = w·x + b` biçimindeki parametreli bir modeli ve hata hesabını anlamak.
- **İki girdili** bir doğrusal model kurmak ve küçük bir aramayla parametreleri seçmek.

> 👨‍🏫 **Hoca notu:** Bu laboratuvar önceki haftanın devamıdır; `print`, değişkenler, `input`, `if`, `for` ve listeler önceden işlendi. Öğrenciler fonksiyon tanımlamayı veya NumPy'ı henüz bilmek zorunda değil. İlk dersin örneklerinde olduğu gibi **anlat → çıktıyı tahmin ettir → çalıştır → öğrenciye uygulat** yaklaşımını kullanın. Kod hücrelerini yukarıdan aşağı sırayla çalıştırın.

**Haftalık planla bağlantı:** 2. hafta teorik başlığı “Parametrik Yöntemler ve Çok Değişkenli Yöntemler”; ön hazırlık “Temel istatistik, olasılık ve doğrusal cebir”; uygulama “Python ile örnek problem çözümü”. Bu dosya konuları sıfırdan, küçük hesaplarla somutlaştırır.

## ⏱️ Ders akışı / hoca kontrol tablosu

| Zaman | Bölüm | Uygulama |
|---|---|---|
| 00–10 dk | 1 · Tekrar | Not listesini inceleme, geçme sayısı |
| 10–25 dk | 2 · Temel istatistik | Ortalama, varyans, standart sapma |
| 25–35 dk | 3 · Olasılık | Gözlenen satın alma olasılığı |
| 35–50 dk | 4 · Doğrusal cebir | Vektör, skaler çarpım, veri matrisi |
| 50–65 dk | 5 · Parametrik model | `y = w*x + b`, tahmin ve hata |
| 65–80 dk | 6 · Çok girdili model | İki özellik ve iki ağırlık |
| 80–90 dk | 7 · Mini proje | Parametre tarama, kapanış |

**Sınıfa sorulacak ana soru:** “Bir modele yalnızca sayıların bulunduğu örnekleri verirsek, hangi işlemlerle yeni bir örnek için tahmin üretebiliriz?”

---
# 1. İlk haftayı hatırlayalım — 10 dk

## Hücre 1.1 · Bir veri listesi

Bir sınıftaki sınav puanları tek tek değişkenlere yazılmak yerine listede tutulabilir. Aşağıdaki kodda geçen haftanın **liste, döngü, karşılaştırma ve sayaç** konularını yeniden kullanıyoruz.

> 👨‍🏫 **Hoca anlatımı:** “Veri kümesi” bazen böyle basit bir listeyle başlar. Henüz model kurmuyoruz; veriyi okuyup özetliyoruz. Önce kodu çalıştırmadan ‘kaç not 70 ve üzeridir?’ diye sorun.

In [1]:
notlar = [45, 70, 85, 60, 90]

toplam = 0
gecen_sayisi = 0

for notu in notlar:
    toplam += notu
    if notu >= 70:
        gecen_sayisi += 1

print("Not sayısı:", len(notlar))
print("Toplam:", toplam)
print("70 ve üzeri:", gecen_sayisi)

Not sayısı: 5
Toplam: 350
70 ve üzeri: 3


**Beklenen sonuç:** 5 not, toplam 350 ve 70 veya üzeri 3 not.

### ✍️ Öğrenci uygulaması 1.2
`notlar` listesindeki **70 altı** notları saydırın. Öğrencilerin önce tek başına denemesini isteyin. Ardından aşağıdaki çözümü açın.

In [2]:
notlar = [45, 70, 85, 60, 90]

kalan_sayisi = 0
for notu in notlar:
    if notu < 70:
        kalan_sayisi += 1

print("70 altı:", kalan_sayisi)

70 altı: 2


👨‍🏫 **Geçiş:** “Veriyi sayabiliyoruz. Şimdi bir veri kümesinin nasıl dağıldığını birkaç sayıyla özetleyelim.”

---
# 2. Temel istatistik — 15 dk

## Hücre 2.1 · Ortalama

**Aritmetik ortalama:** Sayıları topla, gözlem sayısına böl.

\[
\bar{x}=\frac{x_1+x_2+\cdots+x_n}{n}
\]

Veri kümesi: 60, 70, 80, 90, 100. Öğrencilere önce zihinden ortalamayı buldurun.

> 👨‍🏫 **Hoca notu:** `n`, gözlem sayısıdır. Kodda `len()` bu işi yapar. Listede değişiklik yapıldığında hesaplama yine geçerli olur.

In [3]:
puanlar = [60, 70, 80, 90, 100]

toplam = 0
for puan in puanlar:
    toplam += puan

ortalama = toplam / len(puanlar)
print("Ortalama:", ortalama)

Ortalama: 80.0


## Hücre 2.2 · Varyans

Aynı ortalamaya sahip iki grup aynı derecede dağılmış olmayabilir. **Varyans**, değerlerin ortalamadan ne kadar uzaklaştığını ölçmek için sapmaların karelerinin ortalamasını alır.

Bu derste **elimizdeki veri kümesinin tamamı için varyans** hesaplıyoruz: paydada `n` vardır. Örneklem varyansı (paydada `n - 1`) sonraki istatistik derslerinde ayrıca ele alınabilir.

\[
\sigma^2 = \frac{1}{n}\sum_{i=1}^{n}(x_i-\bar{x})^2
\]

**Neden kare?** Pozitif ve negatif sapmaların birbirini götürmesini önler; büyük sapmaları daha güçlü yansıtır.

> 👨‍🏫 **Tahtada çöz:** 60'ın ortalamadan farkı −20, karesi 400. Diğer kareler 100, 0, 100, 400; toplam 1000; 5'e bölünce 200.

In [4]:
puanlar = [60, 70, 80, 90, 100]

ortalama = 80.0  # Önceki hücrede bulduğumuz değer
kareler_toplami = 0

for puan in puanlar:
    sapma = puan - ortalama
    kareler_toplami += sapma ** 2

varyans = kareler_toplami / len(puanlar)
print("Sapmaların kareleri toplamı:", kareler_toplami)
print("Varyans:", varyans)

Sapmaların kareleri toplamı: 1000.0
Varyans: 200.0


## Hücre 2.3 · Standart sapma

**Standart sapma = varyansın karekökü.** Varyans “puan karesi” birimindeyken standart sapma yeniden puan ölçeğine döner.

\[
\sigma = \sqrt{\sigma^2}
\]

Harici kütüphane kullanmadan karekök için `** 0.5` yeterlidir.

In [5]:
standart_sapma = varyans ** 0.5
print("Standart sapma (yaklaşık):", round(standart_sapma, 2))

Standart sapma (yaklaşık): 14.14


### ✍️ Öğrenci uygulaması 2.4 — Çözümlü

Bu kez veri **[10, 20, 30]** olsun. Ortalama, anakütle varyansı ve standart sapmayı kendiniz hesaplayın.

**Önce sınıfa sor:** Ortalama 20 ise hangi sayıların sapma kareleri sıfırdan farklıdır?

**Beklenen:** Ortalama 20; varyans yaklaşık 66,67; standart sapma yaklaşık 8,16.

In [6]:
veriler = [10, 20, 30]

toplam = 0
for deger in veriler:
    toplam += deger
ortalama = toplam / len(veriler)

kareler_toplami = 0
for deger in veriler:
    kareler_toplami += (deger - ortalama) ** 2

varyans = kareler_toplami / len(veriler)
standart_sapma = varyans ** 0.5

print("Ortalama:", ortalama)
print("Varyans:", round(varyans, 2))
print("Standart sapma:", round(standart_sapma, 2))

Ortalama: 20.0
Varyans: 66.67
Standart sapma: 8.16


> 👨‍🏫 **Dikkat:** `round` yalnızca ekranda okunmasını kolaylaştırmak için kullanılır. Hassasiyeti korumak istediğiniz hesaplarda ara sonuçları yuvarlamayın.

---
# 3. Olasılık: Veriden olasılığa — 10 dk

## Hücre 3.1 · Gözlenen sıklık

Bir kırtasiyede kaydedilen **10 satın alma işlemi** için şu ürün etiketlerini görüyoruz:

`["Kalem", "Defter", "Kalem", "Silgi", "Kalem", "Defter", "Kalem", "Silgi", "Defter", "Kalem"]`

Verideki **gözlenen satın alma oranını** hesaplıyoruz:

\[
\widehat{P}(\text{Kalem}) = \frac{\text{Kalem geçen işlem sayısı}}{\text{Tüm işlem sayısı}}
\]

Bu 10 işlemde kalem oranı %50'dir. Bu, her dönemde kalem alma olasılığının mutlaka %50 olduğu anlamına gelmez; elimizdeki veriden elde edilmiş bir **tahmindir**.

> 👨‍🏫 **Sor:** Bu 10 kayıttan rastgele bir kayıt seçsek “Kalem” gelme oranı kaçtır?

In [7]:
satislar = ["Kalem", "Defter", "Kalem", "Silgi", "Kalem",
            "Defter", "Kalem", "Silgi", "Defter", "Kalem"]

kalem_sayisi = 0
for urun in satislar:
    if urun == "Kalem":
        kalem_sayisi += 1

p_kalem = kalem_sayisi / len(satislar)
print("Kalem sayısı:", kalem_sayisi)
print("Gözlenen Kalem oranı:", p_kalem)
print("Yüzde:", p_kalem * 100)

Kalem sayısı: 5
Gözlenen Kalem oranı: 0.5
Yüzde: 50.0


### ✍️ Öğrenci uygulaması 3.2 — Çözümlü

Aynı listede **Defter** ve **Silgi** için gözlenen oranları hesaplayın. Üç oranı toplayın; sonuç neden 1 olmalı?

**Beklenen:** Defter 0,30; Silgi 0,20; toplam 1,00. Çünkü her kayıt tam olarak bu üç üründen birini gösteriyor.

In [8]:
defter_sayisi = 0
silgi_sayisi = 0

for urun in satislar:
    if urun == "Defter":
        defter_sayisi += 1
    elif urun == "Silgi":
        silgi_sayisi += 1

p_defter = defter_sayisi / len(satislar)
p_silgi = silgi_sayisi / len(satislar)

print("Defter:", round(p_defter, 2))
print("Silgi:", round(p_silgi, 2))
print("Toplam:", round(p_kalem + p_defter + p_silgi, 2))

Defter: 0.3
Silgi: 0.2
Toplam: 1.0


> 👨‍🏫 **Geçiş:** “Şimdi tek bir gözleme ait birden fazla ölçüyü birlikte kullanacağız. Bir ev için metrekare ve oda sayısı, bir öğrenci için çalışma saati ve çözdüğü test sayısı gibi...”

---
# 4. Doğrusal cebire giriş: Vektörler ve veri matrisi — 15 dk

## Hücre 4.1 · Vektör, toplama ve sayıyla çarpma

**Vektör:** Belirli sırayla dizilmiş sayılar. Python'da başlangıç için bir listeyle temsil edeceğiz.

Örneğin `[2, 3]` bir öğrencinin `[çalışma saati, çözdüğü test sayısı]` değerleri olabilir. Vektörün ilk elemanını ikinciyle karıştırmamak gerekir.

\[
[2,3]+[4,5]=[6,8]
\]

\[
2\cdot[2,3]=[4,6]
\]

> 👨‍🏫 **Not:** Buradaki çarpma vektörün tüm elemanlarına aynı sayıyı uygulamaktır.

In [9]:
a = [2, 3]
b = [4, 5]

toplam_vektor = []
ikile_carpilmis = []

for i in range(len(a)):
    toplam_vektor.append(a[i] + b[i])
    ikile_carpilmis.append(2 * a[i])

print("Vektör toplamı:", toplam_vektor)
print("İkiyle çarpım:", ikile_carpilmis)

Vektör toplamı: [6, 8]
İkiyle çarpım: [4, 6]


## Hücre 4.2 · Skaler çarpım (dot product)

Aynı uzunluktaki iki vektörün karşılıklı elemanları çarpılıp toplanır; sonuç **tek bir sayı** olur:

\[
[2,3]\cdot[4,5] = 2\cdot4 + 3\cdot5 = 23
\]

Çok girdili doğrusal modellerde **özellikler** ile **ağırlıkları** birleştirmenin temel yoludur.

In [10]:
ozellikler = [2, 3]
agirliklar = [4, 5]

skaler_carpim = 0
for i in range(len(ozellikler)):
    skaler_carpim += ozellikler[i] * agirliklar[i]

print("Skaler çarpım:", skaler_carpim)

Skaler çarpım: 23


## Hücre 4.3 · Matris fikri: Her satır bir gözlem

Birden fazla öğrencinin verisi **liste içinde liste** biçiminde tutulabilir. Gerçek projelerde matris işlemleri için kütüphaneler kullanılır; bugün mantığını çıplak Python'la göreceğiz.

| Öğrenci | Çalışma saati | Test sayısı |
|---|---:|---:|
| 1 | 1 | 2 |
| 2 | 2 | 1 |
| 3 | 3 | 3 |

Bu düzenin şekli **3 satır × 2 özellik**. Sonraki bölümde modele verdiğimiz `x` değerleri bu satırlardan gelecek.

In [11]:
ogrenciler = [
    [1, 2],
    [2, 1],
    [3, 3]
]

print("Öğrenci sayısı:", len(ogrenciler))
print("İkinci öğrencinin verisi:", ogrenciler[1])
print("Üçüncü öğrencinin çalışma saati:", ogrenciler[2][0])

for satir in ogrenciler:
    print("Çalışma saati:", satir[0], "| Test sayısı:", satir[1])

Öğrenci sayısı: 3
İkinci öğrencinin verisi: [2, 1]
Üçüncü öğrencinin çalışma saati: 3
Çalışma saati: 1 | Test sayısı: 2
Çalışma saati: 2 | Test sayısı: 1
Çalışma saati: 3 | Test sayısı: 3


### ✍️ Öğrenci uygulaması 4.4 — Çözümlü

Bir öğrenci `[2, 5]` özelliklerine ve model `[4, 3]` ağırlıklarına sahip olsun. `ağırlıklı toplam + 10` değerini **döngü** ile hesaplayın.

**Beklenen:** `2*4 + 5*3 + 10 = 33`. Bunun henüz eğitim sonucu değil, elimizdeki ağırlıklarla yapılan bir hesap olduğuna dikkat çekin.

In [12]:
ozellikler = [2, 5]
agirliklar = [4, 3]
sabit_terim = 10

sonuc = 0
for i in range(len(ozellikler)):
    sonuc += ozellikler[i] * agirliklar[i]

sonuc += sabit_terim
print("Ağırlıklı sonuç:", sonuc)

Ağırlıklı sonuç: 33


---
# 5. Parametrik yöntem: Tek değişkenli doğrusal model — 15 dk

## Hücre 5.1 · Parametre nedir?

Parametrik yaklaşımda modelin bir **biçimini** seçeriz; eğitim verisi bu biçimin içindeki sayıları belirlemek için kullanılır. Örnek:

\[
\hat y = wx+b
\]

- `x`: Girdi / özellik (ör. çalışma saati).
- `ŷ`: Modelin tahmini (ör. sınav puanı).
- `w`: Girdinin katkısını belirleyen **ağırlık / eğim**.
- `b`: Sabit terim (**bias / kesişim**).
- `w` ve `b`: Modelin **öğrenilecek parametreleri**.

**Örnek veri:** `[x, gerçek not]` olarak `[[1,20],[2,30],[3,40],[4,50]]`.

Bu küçük örnek için `w=10`, `b=10` verileri tam karşılar. **Gerçek hayatta puanlar genellikle bu kadar kusursuz doğrusal değildir.**

> 👨‍🏫 **Önemli fark:** `x` öğrencinin ölçülen bilgisi, `w` ve `b` ise modelin ayarladığı değerlerdir. Öğrencilerden bunları ayrı ayrı söylemelerini isteyin.

In [13]:
calisma_saati = 5
w = 10
b = 10

tahmin = w * calisma_saati + b
print("5 saat için tahmin:", tahmin)

5 saat için tahmin: 60


## Hücre 5.2 · Hata ve ortalama karesel hata (MSE)

Gerçek notlarla modelin tahminleri arasındaki farkı değerlendirmek gerekir:

\[
\mathrm{MSE}=\frac{1}{n}\sum_{i=1}^{n}(y_i-\hat y_i)^2
\]

Burada `y` **gerçek değer**, `ŷ` **tahmin**. Kare, pozitif ve negatif hataların birbirini götürmesini engeller. Küçük MSE, **bu değerlendirilen veri üzerindeki** daha küçük karesel hata demektir.

> 👨‍🏫 **Sor:** “Kendi verimizde sıfır hata bulmak, hiç görmediğimiz öğrencileri de sıfır hatayla tahmin edeceğimiz anlamına gelir mi?” **Hayır**; eğitim/test ayrımı 3. haftanın konusu.

In [14]:
egitim_verisi = [
    [1, 20],
    [2, 30],
    [3, 40],
    [4, 50]
]

w = 10
b = 10
kare_hata_toplami = 0

for satir in egitim_verisi:
    x = satir[0]
    gercek = satir[1]
    tahmin = w * x + b
    kare_hata_toplami += (gercek - tahmin) ** 2
    print("Saat:", x, "Gerçek:", gercek, "Tahmin:", tahmin)

mse = kare_hata_toplami / len(egitim_verisi)
print("MSE:", mse)

Saat: 1 Gerçek: 20 Tahmin: 20
Saat: 2 Gerçek: 30 Tahmin: 30
Saat: 3 Gerçek: 40 Tahmin: 40
Saat: 4 Gerçek: 50 Tahmin: 50
MSE: 0.0


### ✍️ Öğrenci uygulaması 5.3 — Çözümlü

Aynı veride `w=8`, `b=10` deneyin. Tahminleri ve MSE'yi hesaplayın. Model önceki kadar iyi uyuyor mu?

**Beklenen:** Tahminler 18, 26, 34, 42; kare hatalar 4, 16, 36, 64; MSE 30.0.

In [15]:
egitim_verisi = [[1, 20], [2, 30], [3, 40], [4, 50]]
w = 8
b = 10
kare_hata_toplami = 0

for satir in egitim_verisi:
    x = satir[0]
    gercek = satir[1]
    tahmin = w * x + b
    kare_hata_toplami += (gercek - tahmin) ** 2
    print("Tahmin:", tahmin, "Kare hata:", (gercek - tahmin) ** 2)

mse = kare_hata_toplami / len(egitim_verisi)
print("MSE:", mse)

Tahmin: 18 Kare hata: 4
Tahmin: 26 Kare hata: 16
Tahmin: 34 Kare hata: 36
Tahmin: 42 Kare hata: 64
MSE: 30.0


---
# 6. Çok girdili doğrusal model — 15 dk

## Hücre 6.1 · Bir girdiden iki girdiye

Tek bir çalışma saati çoğu zaman yeterli olmayabilir. Bu kez:

- `x1`: Çalışma saati
- `x2`: Çözülen deneme sayısı
- `ŷ`: Tahmin edilen puan

\[
\hat y=b+w_1x_1+w_2x_2
\]

Bugünkü oyuncak modelde `b = 10`, `w1 = 5`, `w2 = 2`:

\[
\hat y=10+5\times\text{çalışma saati}+2\times\text{deneme sayısı}
\]

Bu, **birden fazla girdi (çok özellikli) kullanan doğrusal model** örneğidir. Terminoloji notu: Bazı kaynaklar “çok değişkenli” ifadesini *birden fazla çıktı* için de kullanır; burada dersin temel düzeyine uygun biçimde **iki giriş değişkenini** inceliyoruz.

> 👨‍🏫 **Örnek:** 2 saat çalışan, 5 deneme çözen öğrenci için tahmin = `10 + 5*2 + 2*5 = 30`. Parametreleri şimdilik *biz seçtik*; birazdan nasıl seçilebileceklerini göstereceğiz.

In [16]:
calisma_saati = 2
deneme_sayisi = 5

b = 10
w1 = 5
w2 = 2

tahmin = b + w1 * calisma_saati + w2 * deneme_sayisi
print("Tahmin edilen puan:", tahmin)

Tahmin edilen puan: 30


## Hücre 6.2 · İki özellikli eğitim verisi

Bu sınıf içi örnek için **özellikle basit ve yapay** dört gözlem oluşturulmuştur. Gerçek sınav verisi olduğunu düşünmeyin.

| Saat | Deneme | Gerçek puan |
|---:|---:|---:|
| 1 | 2 | 19 |
| 2 | 1 | 22 |
| 3 | 3 | 31 |
| 4 | 2 | 34 |

Her satır `[x1, x2, y]` biçiminde saklanır. Aynı formülle her satırı tahmin edeceğiz.

In [17]:
veri = [
    [1, 2, 19],
    [2, 1, 22],
    [3, 3, 31],
    [4, 2, 34]
]

b = 10
w1 = 5
w2 = 2
kare_hata_toplami = 0

for satir in veri:
    x1 = satir[0]
    x2 = satir[1]
    gercek = satir[2]
    tahmin = b + w1 * x1 + w2 * x2
    kare_hata_toplami += (gercek - tahmin) ** 2
    print(f"Saat={x1}, Deneme={x2}, Gerçek={gercek}, Tahmin={tahmin}")

print("MSE:", kare_hata_toplami / len(veri))

Saat=1, Deneme=2, Gerçek=19, Tahmin=19
Saat=2, Deneme=1, Gerçek=22, Tahmin=22
Saat=3, Deneme=3, Gerçek=31, Tahmin=31
Saat=4, Deneme=2, Gerçek=34, Tahmin=34
MSE: 0.0


## Hücre 6.3 · Aynı hesabın vektör karşılığı

İki girdi ve iki ağırlık varsa: `b + (özellikler · ağırlıklar)`.

\[
\hat{y}=b+\mathbf{w}^{T}\mathbf{x}
\]

> 👨‍🏫 **Soru:** Yukarıdaki modelde ağırlıklar `[5, 2]` ise öğrenci verisi `[2, 5]` için sonuç ne olur?

In [18]:
ozellikler = [2, 5]
agirliklar = [5, 2]
b = 10

tahmin = b
for i in range(len(ozellikler)):
    tahmin += ozellikler[i] * agirliklar[i]

print("Skaler çarpımla tahmin:", tahmin)

Skaler çarpımla tahmin: 30


### ✍️ Öğrenci uygulaması 6.4 — Çözümlü

Aynı model (`b=10`, `w1=5`, `w2=2`) ile **5 saat ve 4 deneme** için tahmin üretin. Hem açık formülle hem vektör döngüsüyle aynı sonucu bulun.

**Beklenen:** `10 + 5*5 + 2*4 = 43`. **Dikkat:** Bu örnek eğitim verilerinin dışındaki bir girdidir; gerçek puanı bilmiyoruz.

In [19]:
b = 10
w1 = 5
w2 = 2
x1 = 5
x2 = 4

acik_tahmin = b + w1 * x1 + w2 * x2

ozellikler = [x1, x2]
agirliklar = [w1, w2]
vektor_tahmin = b
for i in range(len(ozellikler)):
    vektor_tahmin += ozellikler[i] * agirliklar[i]

print("Açık formül:", acik_tahmin)
print("Vektörle:", vektor_tahmin)

Açık formül: 43
Vektörle: 43


---
# 7. Mini proje: Parametreleri veriye göre seç — 10 dk

## Hücre 7.1 · Problem ve öğrenciye verilecek kısa görev

Şimdi `b=10, w1=5, w2=2` değerlerini kendimiz seçmek yerine, **küçük bir aday kümesini deneyerek** bulacağız.

- Dört öğrenci kaydı aynı kalsın.
- `b` için 0–15, `w1` için 0–7 ve `w2` için 0–4 aralığındaki **tam sayıları** deneyelim.
- Her üçlü için dört kayıt üzerinde MSE hesaplayalım.
- En küçük MSE'yi veren üçlüyü saklayalım.

**Öğrencilere görev:** Aşağıdaki çözümü göstermeden önce şu üç boşluğu sor: “Hata toplama değişkenini **hangi döngünün başında** sıfırlarız?”, “En iyi modeli ne zaman güncelleriz?”, “En iyi model sıfır hatalıysa ne söyleyebiliriz?”

> 👨‍🏫 **Terminoloji:** Bu bir **ızgara taraması (grid search) ile parametre seçimi** gösterimidir; yaygın doğrusal regresyon eğitiminin tek veya en verimli yolu değildir. Gerçek veride sürekli ağırlıklar ve başka optimizasyon yöntemleri de kullanılır. Burada kolay izlenebilmesi için küçük tam sayı aralığı taranıyor.

In [20]:
# MİNİ PROJE — TAM HOCA ÇÖZÜMÜ
veri = [
    [1, 2, 19],
    [2, 1, 22],
    [3, 3, 31],
    [4, 2, 34]
]

# Henüz adayları değerlendirmedik.
en_iyi_mse = float("inf")
en_iyi_b = None
en_iyi_w1 = None
en_iyi_w2 = None

deneme_sayisi = 0

for b in range(0, 16):
    for w1 in range(0, 8):
        for w2 in range(0, 5):
            deneme_sayisi += 1
            kare_hata_toplami = 0

            for satir in veri:
                x1 = satir[0]
                x2 = satir[1]
                gercek = satir[2]
                tahmin = b + w1 * x1 + w2 * x2
                kare_hata_toplami += (gercek - tahmin) ** 2

            mse = kare_hata_toplami / len(veri)

            if mse < en_iyi_mse:
                en_iyi_mse = mse
                en_iyi_b = b
                en_iyi_w1 = w1
                en_iyi_w2 = w2

print("Denenen parametre üçlüsü:", deneme_sayisi)
print("Bulunan b, w1, w2:", en_iyi_b, en_iyi_w1, en_iyi_w2)
print("Eğitim verisindeki MSE:", en_iyi_mse)

Denenen parametre üçlüsü: 640
Bulunan b, w1, w2: 10 5 2
Eğitim verisindeki MSE: 0.0


**Beklenen çıktı:** 640 aday; `b=10`, `w1=5`, `w2=2`, eğitim verisi üzerinde `MSE=0.0`.

**Bu sonucun yorumu:** Yapay olarak tam doğru modelden üretilmiş dört kayıt ve doğru parametrelerin bulunduğu aday aralığı kullanıldı. Buradaki sıfır hata, *gerçek hayatta* veya *hiç görülmeyen örneklerde* sıfır hata garantisi değildir.

### ✍️ Öğrenci uygulaması 7.2 — Bulduğun modeli kullan

Mini projede bulunan parametreleri kullanarak `[çalışma saati=5, deneme sayısı=4]` için tahmin üretin.

**Beklenen:** 43.

In [21]:
yeni_saat = 5
yeni_deneme = 4

yeni_tahmin = (en_iyi_b
               + en_iyi_w1 * yeni_saat
               + en_iyi_w2 * yeni_deneme)

print("Yeni öğrenci için tahmin:", yeni_tahmin)

Yeni öğrenci için tahmin: 43


---
# Ders bitişi — 2 dakikalık sözlü kontrol

**1. Soru:** Özellik `x` ile parametre `w` arasındaki fark nedir?  
**Yanıt:** `x` gözlemin ölçülen girdisi; `w` modelin bu girdiye verdiği ağırlıktır.

**2. Soru:** Ortalama ile varyans aynı şeyi mi anlatır?  
**Yanıt:** Hayır. Ortalama merkezî düzeyi; varyans etrafa yayılımı anlatır.

**3. Soru:** `[2,3] · [4,5]` kaçtır?  
**Yanıt:** 23.

**4. Soru:** MSE neden karesel hataları toplar?  
**Yanıt:** Pozitif/negatif hatalar birbirini götürmesin diye; büyük hatalara daha fazla ağırlık verir.

**5. Soru:** Bu örnekteki sıfır eğitim hatası yeni veri için garanti midir?  
**Yanıt:** Hayır. Bir sonraki hafta eğitim-test ayrımı ve performans değerlendirmeye geçeceğiz.

### 75 dakikalık uygulama seçeneği
- İlk hafta tekrarını 5 dakikaya indir.
- Olasılıktaki ikinci uygulamayı ev çalışması yap.
- Vektör toplamı için yalnızca bir örnek göster; skaler çarpımı mutlaka işle.
- Grid search çözümünün iç içe döngülerini adım adım birlikte yazmak yerine hazır hücreyle göster ve değişkenlerin rollerini sorgulat.

### Eğitmen için yöntemsel sınırlar
- Bu notebook **örneklem varyansını** değil, n'ye bölünen anakütle varyansını hesaplar.
- Olasılık bölümü, eldeki örneklemdeki **ampirik oranları** hesaplar; bir topluma genelleme yapmak için daha fazla varsayım ve veri gerekir.
- “Çok değişkenli” başlığında **iki girdili, tek çıktılı** bir modelin matematiği işlenmiştir.
- Modelin **eğitim verisindeki MSE'si** hesaplanmıştır. Eğitim-test ayrımı, çapraz doğrulama ve model karşılaştırması ders planında **3. haftanın** konusudur.
- Gerçek dünyada sınav puanı sınırsız doğrusal olmayabilir; bu örnek, parametre ve özellik fikrini göstermek için bilinçli olarak basitleştirilmiştir.